IconThemePlugin.js

```diff
@@ -0,0 +1,37 @@
+class IconThemePlugin {
+  // Define `apply` as its prototype method which is supplied with compiler as its argument
+  apply(compiler) {
+    const pluginName = IconThemePlugin.name;
+    const { webpack } = compiler;
+    const { RawSource } = webpack.sources;
+
+    // Specify the event hook to attach to
+    compiler.hooks.emit.tap(pluginName, (compilation) => {
+      const icons = compilation.getAssets().map(
+        (asset) => asset.name.split('/').pop()
+      );
+
+      // Create an object with the property name as the asset name
+      // and the value as the file extension
+      const iconTypes = icons.reduce(
```

Comment:
I find that `reduce` adds quite a bit of cognitive load.

You can write this as the following:

```js
const iconEntries = icons.map(icon => icon.split('.')) // NB: This does not work as you'd expect if someone puts dots in their names (same with your original implementation):)
const iconTypes = Object.fromEntries(iconEntries)
```

Reply:
I can easily mod the `reduce` to use `pop()` on the split to get the extension:

```js
const iconTypes = icons.reduce(
	(acc, icon) => {
		const split = icon.split('.');
		const ext = split.pop();
		const name = split.join('.');
		acc[name] = ext;
		return acc;
	},
	{}
);
```

But I'm not sure if there is an easy way to do so with `map`. Any suggestions?

Reply:
Suppose it could be done like this:

```js
const iconEntries = icons.map(file => [
  file.substr(0,file.lastIndexOf('.')),
  file.substr(file.lastIndexOf('.')+1, file.length)
])

const iconTypes = Object.fromEntries(iconEntries)
```

Reply:
Whoa I had no idea that `Object.fromEntries` would work with an array of arrays like that, cool stuff! Good idea.